Source/x11/XGServerWindow.m

```diff
@@ -4409,75 +4413,144 @@ - (void) freecursor: (void*) cid
   return depth;
 }
 
+/* This method assumes that we deal with one X11 screen - `defScreen`.
+   Basically it means that we have DISPLAY variable set to `:0.0`.
+   AppKit and X11 use the same term "screen" with different meaning:
+   AppKit Screen - single mintor/display device.
+   X11 Screen - its a plane where X Server can draw on.
+   XRandR - Xorg extension that halps to manipulate monitors layout providing
+   X11 Screen. */
 - (NSArray *)screenList
 {
-  /* I guess screen numbers are in order starting from zero, but we
-     put the main screen first */
- int i;
- int count = ScreenCount(dpy);
- NSMutableArray *screens = [NSMutableArray arrayWithCapacity: count];
- if (count > 0)
-   [screens addObject: [NSNumber numberWithInt: defScreen]];
- for (i = 0; i < count; i++)
-   {
-     if (i != defScreen)
-       [screens addObject: [NSNumber numberWithInt: i]];
-   }
- return screens;
+  NSMutableArray      *screens;
+  int                 i;
+
+  monitorsCount = 0;
+  if (monitors != NULL) {
+    NSZoneFree([self zone], monitors);
+  }
+  
+#ifdef HAVE_XRANDR
+  XRRScreenResources  *screen_res;
+  XRROutputInfo       *output_info;
+  XRRCrtcInfo         *crtc_info;
+
+  screen_res = XRRGetScreenResources(dpy, RootWindow(dpy, defScreen));
+  monitorsCount = screen_res->noutput;
+
+  if (monitorsCount > 0)
+    {
+      screens = [NSMutableArray arrayWithCapacity: monitorsCount];
+      monitors = NSZoneMalloc([self zone], monitorsCount * sizeof(MonitorDevice));
+
+      for (i = 0; i < monitorsCount; i++)
+        {
+          output_info = XRRGetOutputInfo(dpy, screen_res, screen_res->outputs[i]);
+          if (output_info->crtc)
+            {
+              crtc_info = XRRGetCrtcInfo(dpy, screen_res, output_info->crtc);
+
+              // Fill the cache of device parameters
+              monitors[i].depth = [self windowDepthForScreen: i];
```

Comment:
Why don't you pass defScreen here? That way the called methods could stay the same.
This will anyway result in all you monitors having the same depth and resolution, which seems wrong to me.

Reply:
Good, done.